src/java.base/share/classes/java/text/DecimalFormat.java

```diff
@@ -1790,22 +1782,23 @@ String fastFormat(double d) {
     }
 
     /**
-     * Sets the {@code DigitList} used by this {@code DecimalFormat}
+     * Utility method that sets the {@code DigitList} used by this {@code DecimalFormat}
      * instance.
+     *
      * @param number the number to format
      * @param isNegative true, if the number is negative; false otherwise
      * @param maxDigits the max digits
+     * @throws AssertionError if provided a Number subclass that is not supported
+     *         by {@code DigitList}
      */
     void setDigitList(Number number, boolean isNegative, int maxDigits) {
-
-        if (number instanceof Double) {
-            digitList.set(isNegative, (Double) number, maxDigits, true);
-        } else if (number instanceof BigDecimal) {
-            digitList.set(isNegative, (BigDecimal) number, maxDigits, true);
-        } else if (number instanceof Long) {
-            digitList.set(isNegative, (Long) number, maxDigits);
-        } else if (number instanceof BigInteger) {
-            digitList.set(isNegative, (BigInteger) number, maxDigits);
+        switch (number) {
+            case Double d -> digitList.set(isNegative, d, maxDigits, true);
+            case BigDecimal bd -> digitList.set(isNegative, bd, maxDigits, true);
+            case Long l -> digitList.set(isNegative, l, maxDigits);
+            case BigInteger bi -> digitList.set(isNegative, bi, maxDigits);
+            default -> throw new AssertionError(
+                    String.format("DigitList does not support %s", number.getClass().getName()));
```

Comment:
Nit: could be `"DigitList does not support %s".formatted(number.getClass().getName())`